plugins/modules/foreman_ansible_roles_facts.py

```diff
@@ -0,0 +1,186 @@
+    type: str
+  proxy:
+    description:
+      - The Smart Porxy name to use for fetching Ansible Roles
+    required: false
+    type: str
+    alias: smart_proxy
+  search:
+    description:
+      - Search query to use
+      - If None, all resources are returned
+    type: str
+  full_details:
+    description:
+      - If C(True) all details about the found resources are returned
+    type: bool
+    default: false
+    aliases: [ info ]
+  organization:
+    description:
+      - Organization that the role is in
+    required: false
+    type: str
+  location:
+    description:
+      - Location that the role is in
+    required: false
+    type: str
+extends_documentation_fragment:
+  - foreman
+'''
+
+
+EXAMPLES = '''
+- name: Get Ansible Roles available to import on a proxy
+  foreman_ansible_roles:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    state: "fetch"
+    proxy: "foreman.example.com"
+  register: result
+- debug:
+    var: result
+
+- name: Get all imported Ansible Roles
+  foreman_ansible_roles:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    organization: "
+    state: "list"
+  register: result
+- debug:
+    var: result
+
+- name: Find the imported Ansible Role named example_role
+  foreman_ansible_roles:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    organization: "
+    state: "list"
+    search: "name=example_role"
+  register: result
+- debug:
+    var: result
+'''
+
+RETURN = '''
+ansible_roles:
+  description: Roles that are able to be imported on the proxy
+  returned: state = fetched
+  type: list
+'''
+
+from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
+
+
+class ForemanAnsibleRolesFactsModule(ForemanEntityAnsibleModule):
+    pass
+
+
+def main():
+
+    module = ForemanAnsibleRolesFactsModule(
+        foreman_spec = dict(
+            action = dict(default='fetch', choices=['fetch', 'list', 'show']),
+            proxy = dict(type='entity', flat_name='proxy_id', aliases=['smart_proxy'], resource_type='smart_proxies'),
+            search=dict(default=""),
+            full_details=dict(type='bool', aliases=['info'], default='false'),
+            organization=dict(type='entity', flat_name='organization_id', resource_type='organizations'),
+            location=dict(type='entity', flat_name='location_id', resource_type='locations'),
+        ),
+
+        required_if = [
+            ['action', 'fetch', ['proxy']]
+        ],
+    )
+
+    module_params = module.clean_params()
+    resource = 'ansible_roles'
+    search = module_params['search']
+    params = {}
+
+    if module_params['action'] == 'show':
+        module_params['action'] = 'list'
+        module_params['full_details'] = True
+
+    with module.api_connection():
+        _entity, module_params = module.resolve_entities(module_params, 'ansible_roles')
```

Comment:
```suggestion
        module.auto_lookup_entities()
        params = _flatten_entity(module.foreman_params, module.foreman_spec)
```

I'm afraid we managed to leave this PR around for too long and merged a huge refactoring since then. I'll provide suggestions how to update.

Reply:
Fixed.